### PR TITLE
Redundancy and heartbeat integration

### PR DESCRIPTION
**Pull request by iulianpascalau**
- made the heartbeat sender to use another BLS key when the node is running in redundancy mode with the main node activated

### CODE REVIEW

File: redundancy/redundancy.go [hunk 1]
```diff
@@ -20,17 +21,29 @@ type nodeRedundancy struct {
 	roundsOfInactivity  uint64
 	mutNodeRedundancy   sync.RWMutex
 	messenger           P2PMessenger
+	observerPrivateKey  crypto.PrivateKey
+}
+
+// ArgNodeRedundancy represents the DTO structure used nodeRedundancy constructor
```
Comment (SebastianMarian) [inline redundancy/redundancy.go:27]: used by or used in ?

File: heartbeat/interface.go [hunk 2]
```diff
@@ -112,3 +113,11 @@ type CurrentBlockProvider interface {
 	GetCurrentBlockHeader() data.HeaderHandler
 	IsInterfaceNil() bool
 }
+
+// NodeRedundancyHandler defines
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply heartbeat/interface.go:117]: done

File: cmd/node/main.go [hunk 3]
```diff
@@ -1373,7 +1373,21 @@ func startNode(ctx *cli.Context, log logger.Logger, version string) error {
 		processComponents.TxLogsProcessor.EnableLogToBeSavedInCache()
 	}
 
-	nodeRedundancy, err := redundancy.NewNodeRedundancy(preferencesConfig.Preferences.RedundancyLevel, networkComponents.NetMessenger)
+	observerBLSPrivateKey, observerBLSPublicKey := cryptoComponents.BlockSignKeyGen.GeneratePair()
+	observerBLSPublicKeyBuff, err := observerBLSPublicKey.ToByteArray()
+	if err != nil {
+		log.Error("error generating observerBLSPublicKeyBuff", "error", err)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply cmd/node/main.go:1379]: No need as the observerBLSPublicKeyBuff (that can be nil) is used only in logs. The argument takes the observerBLSPrivateKey as crypto.PrivateKey. So, for the sake of a log line, I think we can keep it as it is.

File: heartbeat/process/sender.go [hunk 4]
```diff
@@ -87,6 +94,8 @@ func NewSender(arg ArgHeartbeatSender) (*Sender, error) {
 		peerMessenger:        arg.PeerMessenger,
 		peerSignatureHandler: arg.PeerSignatureHandler,
 		privKey:              arg.PrivKey,
+		publicKey:            arg.PrivKey.GeneratePublic(),
+		observerPublicKey:    arg.RedundancyHandler.ObserverPrivateKey().GeneratePublic(),
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply heartbeat/process/sender.go:98]: Right, changed the code + added unit test for it.

File: heartbeat/process/sender_test.go [hunk 5]
```diff
@@ -292,6 +305,202 @@ func TestSender_SendHeartbeatShouldWork(t *testing.T) {
...
+	pubKey := &mock.PublicKeyMock{
+		ToByteArrayHandler: func() (i []byte, e error) {
+			return pkBytes, nil
+		},
+	}
+	signature := []byte("signature")
+
+	broadcastCalled := false
+	signCalled := false
+	genPubKeyClled := false
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply heartbeat/process/sender_test.go:322]: fixed

File: heartbeat/process/sender_test.go [hunk 6]
```diff
@@ -292,6 +305,202 @@ func TestSender_SendHeartbeatShouldWork(t *testing.T) {
...
+	pubKey := &mock.PublicKeyMock{
+		ToByteArrayHandler: func() (i []byte, e error) {
+			return originalPkBytes, nil
+		},
+	}
+	signature := []byte("signature")
+
+	broadcastCalled := false
+	signCalled := false
+	genPubKeyClled := false
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply heartbeat/process/sender_test.go:375]: fixed all occurrences

**Review by SebastianMarian — APPROVED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

**Review by valentincraciun99 — APPROVED** [review]
system testing passed

**Pull request merged**